File: agent/watchlist/service.py
from __future__ import annotations

import inspect
import json
import re
import time
from collections.abc import Awaitable, Callable
from pathlib import Path
from typing import Any

from loguru import logger

from ..model_router import ModelRouter
from ..telemetry import TokenTracker
from .models import WatchlistDecision
from .store import WatchlistStore

DecisionFn = Callable[[str, list[str]], WatchlistDecision | Awaitable[WatchlistDecision]]


class WatchlistService:
    def __init__(
        self,
        root: Path,
        *,
        model_router: ModelRouter | None = None,
        token_tracker: TokenTracker | None = None,
        decider: DecisionFn | None = None,
    ):
        self.root = Path(root).resolve()
        self.store = WatchlistStore(self.root)
        self.model_router = model_router
        self.token_tracker = token_tracker
        self.decider = decider

    def payload(self) -> dict[str, Any]:
        return self.store.payload()

    def read(self) -> str:
        return self.store.read()

    def write(self, content: str) -> dict[str, Any]:
        self.store.write(content)
        return self.payload()

    async def check(self) -> WatchlistDecision:
        content = self.store.read()
        items = self.store.active_items()
        if not items:
            decision = WatchlistDecision.skip("watchlist has no active items")
            self.store.write_decision(decision)
            return decision
        if self.decider:
            result = self.decider(content, items)
            decision = await result if inspect.isawaitable(result) else result
        else:
            decision = await self._decide_with_model(content, items)
        decision.checked_at = decision.checked_at or time.time()
        self.store.write_decision(decision)
        return decision

    async def _decide_with_model(self, content: str, items: list[str]) -> WatchlistDecision:
        if self.model_router is None:
            return WatchlistDecision.skip("model router is unavailable")
        route = self.model_router.route("watchlist_check", task=content)
        snapshot = route.snapshot
        prompt = _decision_prompt(content=content, items=items)
        try:
            resp = await snapshot.provider.chat(
                model=snapshot.model,
                max_tokens=min(1200, snapshot.generation.max_tokens),
                temperature=0,
                reasoning_effort=snapshot.generation.reasoning_effort,
                messages=prompt,
                tools=None,
            )
        except Exception as exc:
            fallback = route.fallback
            if fallback is None:
                raise
            logger.warning("watchlist secondary fallback: {} -> {} because {}", snapshot.model, fallback.model, exc)
            snapshot = fallback
            resp = await fallback.provider.chat(
                model=fallback.model,
                max_tokens=min(1200, fallback.generation.max_tokens),
                temperature=0,
                reasoning_effort=fallback.generation.reasoning_effort,
                messages=prompt,
                tools=None,
            )
        if self.token_tracker and resp.usage:
            self.token_tracker.record(
                snapshot.model,
                resp.usage,
                provider=snapshot.provider_name,
                usage_type="watchlist_check",
                model_role=snapshot.model_role,
            )
        decision = _parse_decision(resp.content or "")
        decision.model = snapshot.model
        decision.provider = snapshot.provider_name
        decision.model_role = snapshot.model_role
        return decision


def _decision_prompt(*, content: str, items: list[str]) -> list[dict[str, str]]:
    return [
        {
            "role": "system",
            "content": (
                "You are a local watchlist decision filter. Decide if the agent should proactively run now. "
                "Return strict JSON only: {\"action\":\"skip|run\",\"reason\":\"...\",\"message\":\"...\"}. "
                "Choose skip unless there is a concrete, timely, user-relevant action. Never include hidden reasoning."
            ),
        },
        {
            "role": "user",
            "content": (
                "Current watchlist markdown:\n"
                f"{content}\n\n"
                "Active items:\n"
                + "\n".join(f"- {item}" for item in items)
            ),
        },
    ]


def _parse_decision(text: str) -> WatchlistDecision:
    raw = text.strip()
    match = re.search(r"\{.*\}", raw, flags=re.DOTALL)
    if match:
        raw = match.group(0)
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        return WatchlistDecision.skip("watchlist model returned non-JSON decision")
    if not isinstance(parsed, dict):
        return WatchlistDecision.skip("watchlist model returned invalid decision")
    decision = WatchlistDecision.from_dict(parsed)
    decision.reason = _clean(decision.reason)[:500]
    decision.message = _clean(decision.message)[:1200]
    if decision.action == "run" and not decision.message:
        decision.action = "skip"
        decision.reason = decision.reason or "run decision had no actionable message"
    return decision


def _clean(value: str) -> str:
    return " ".join(str(value or "").split())

File: packages/core/src/watchlist/service.ts
import { nowTs } from '../util/time'
import type { TokenTracker } from '../memory/token-tracker'
import type { ModelRouter, ProviderSnapshot } from '../model/router'
import type { LLMResponse } from '../providers/base'
import {
  decisionPrompt,
  parseWatchlistDecision,
  WatchlistDecision,
} from './models'
import { WatchlistStore } from './store'

export type WatchlistDecisionFn = (
  content: string,
  items: string[],
) => WatchlistDecision | Promise<WatchlistDecision>

export class WatchlistService {
  readonly root: string
  readonly store: WatchlistStore
  decider: WatchlistDecisionFn | null
  modelRouter: ModelRouter | null
  tokenTracker: TokenTracker | null

  constructor(
    root: string,
    opts: {
      decider?: WatchlistDecisionFn | null
      modelRouter?: ModelRouter | null
      tokenTracker?: TokenTracker | null
    } = {},
  ) {
    this.root = root
    this.store = new WatchlistStore(root)
    this.decider = opts.decider ?? null
    this.modelRouter = opts.modelRouter ?? null
    this.tokenTracker = opts.tokenTracker ?? null
  }

  payload(): Record<string, unknown> {
    return this.store.payload()
  }
  read(): string {
    return this.store.read()
  }
  write(content: string): Record<string, unknown> {
    this.store.write(content)
    return this.payload()
  }

  async check(): Promise<WatchlistDecision> {
    const content = this.store.read()
    const items = this.store.activeItems()
    if (!items.length) {
      const decision = WatchlistDecision.skip('watchlist has no active items')
      this.store.writeDecision(decision)
      return decision
    }
    const decision = this.decider
      ? await this.decider(content, items)
      : await this.decideWithModel(content, items)
    decision.checked_at = decision.checked_at || nowTs()
    this.store.writeDecision(decision)
    return decision
  }

  private async decideWithModel(
    content: string,
    items: string[],
  ): Promise<WatchlistDecision> {
    if (!this.modelRouter)
      return WatchlistDecision.skip('model router is unavailable')
    const route = this.modelRouter.route('watchlist_check', undefined, content)
    let snapshot = route.snapshot
    let usedFallback = false
    let fallbackReason = ''
    let resp: LLMResponse
    try {
      resp = await callSnapshot(snapshot, content, items)
    } catch (error) {
      if (!route.fallback) throw error
      fallbackReason = error instanceof Error ? error.message : String(error)
      snapshot = route.fallback
      usedFallback = true
      resp = await callSnapshot(snapshot, content, items)
    }
    this.tokenTracker?.record(snapshot.model, resp.usage, {
      provider: snapshot.providerName,
      usageType: 'watchlist_check',
      modelRole: snapshot.modelRole,
      routeReason: snapshot.routeReason,
      usedFallback,
      fallbackReason,
    })
    const decision = parseWatchlistDecision(resp.content || '')
    decision.model = snapshot.model
    decision.provider = snapshot.providerName
    decision.model_role = snapshot.modelRole
    return decision
  }
}

async function callSnapshot(
  snapshot: ProviderSnapshot,
  content: string,
  items: string[],
): Promise<LLMResponse> {
  return snapshot.provider.chat({
    model: snapshot.model,
    maxTokens: Math.min(1200, snapshot.generation.maxTokens),
    temperature: 0,
    reasoningEffort: snapshot.generation.reasoningEffort,
    messages: decisionPrompt({ content, items }),
    tools: null,
  })
}
